### PR TITLE
Nginx template designed for Unicore websites

### PR DESCRIPTION
**Pull request by JayH5**
We're currently using the template used in Unicore. This is good for hosting websites as it exposes each Marathon app with a `domain` label by using the label value for the `server_name` field in nginx.

This is not so good for a "micro-services" system where apps need to talk to each other internally and use nginx to route between one another. In that case, it is helpful to be able to match an nginx `location` rather than `server_name`

### CODE REVIEW

**Comment by JayH5** [comment]
@smn: Ready for review.
We need to think about this and design it properly at some point but for now I want to get things going for the mama-ng work.

What this does is:
- If the app has a label `domain`: the existing Unicore template is used and things are routed to the app based on the domain.
- If the app has a label `location`: the Vumi Go template is used and things are routed to the app based on the URL path.
- If the app has both labels: they'll get added to both templates and I don't know what that will mean for the app.

**Comment by smn** [comment]
I was hoping we could get internal services to just discover each other via Consul directly instead of Nginx. Nginx isn't able to proxy generic TCP connections (HAProxy is though). Do we really need to do this? Do we perhaps need to route internal services via HAProxy (and there also have the option of limiting / preventing access?)

File: etc/consular/nginx-services.ctmpl [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+{{range services}}{{$labels := ls (print "consular/" .Name) | explode}}{{if $labels.location}}
+upstream {{.Name}} { {{range service .Name}}
+    server {{.Address}}:{{.Port}};{{end}}
+}
+{{end}}{{else}}
+# Skipped service {{.Name}} as it does not have a KV domain entry.
+{{end}}
+server {
+    listen 80;
```
Comment (smn) [inline etc/consular/nginx-services.ctmpl:18]: We probably want to mark this as [`internal`](http://nginx.org/en/docs/http/ngx_http_core_module.html#internal)

File: etc/consular/nginx-services.ctmpl [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    {{range services}}{{$labels := ls (print "consular/" .Name) | explode}}{{if $labels.location}}
+    location {{$labels.location | parseJSON}} {
+        proxy_pass http://{{.Name}};
+        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+        proxy_set_header Host $http_host;{{if $labels.internal}}{{if $labels.internal | parseJSON}}
+        internal;{{end}}{{end}}
```
Comment (JayH5) [inline etc/consular/nginx-services.ctmpl:24]: It's optional. See here.

**Comment by JayH5** [comment]
@smn: @jerith and I have looked at various ways around this and there are basically 2 reasons we can't rely on Consul for routing at the moment:
1. Since ports are dynamic with Marathon/Mesos, you need both host and port information to access a service. Most software is not coded to use DNS to get both host and port (SRV records), neither is it coded to ping Consul's HTTP API to get host/port.
2. Even if stuff did get SRV records from Consul DNS, we don't know if/when/where that information will be cached...which could break things when hosts/ports suddenly change.

The reason we're using nginx instead of HAProxy is that nginx has smooth config reloads with no dropped connections. I would be very keen on getting HAProxy to do this job as it is better suited to it but getting HAProxy to reload nicely is quite tricky.

**Comment by JayH5** [comment]
This is another reason I wasn't 100% sold on Consul initially - I'm not actually convinced we need a local DNS server. (The other features _are_ really useful, though).

For the Vumi Go stuff we were really only using DNS to look up fixed (non-containerized, init-launched) services like Redis and Postgres. So those services could be hit at `redis.service.consul`, etc. Which is a nice to have and makes configuration marginally easier, but is not strictly necessary.

File: etc/consular/nginx-services.ctmpl (hunk 2 above)
Comment (smn) [inline etc/consular/nginx-services.ctmpl:24]: Ah :) yay!

**Comment by smn** [comment]
You're right, perhaps we don't need internal DNS at the moment but I don't want to pin services to dns names of the workers either and making everything route on a location path or domain in Nginx is a fairly big assumption about what kinds of services we're going to be running and how these would interact with other things. The current approach rules out anything that doesn't talk HTTP which isn't something I feel we should be assuming at this point. 

About HAProxy, I think having HAProxy drop connections during a restart can be seen as benefit to us at this point and as we are designing this system. We have things that fall over horribly in Vumi when postgresql restarts simply because that particular piece was designed with the assumption that stuff can get restarted at any point in time. The reality is that that happens all the time and clients should handle those situations.

**Comment by smn** [comment]
Hm did some reading, looks like nginx is now capable of proxying generic TCP connections http://nginx.org/en/docs/stream/ngx_stream_core_module.html#stream and https://www.nginx.com/resources/admin-guide/tcp-load-balancing/. Not sure if it's available in the versions we run though.

**Comment by smn** [comment]
Hrm I may be wrong, that may be only available for the commercial offering.

**Comment by jerith** [comment]
> About HAProxy, I think having HAProxy drop connections during a restart can be seen as benefit to us at this point and as we are designing this system.

I disagree. There's a lot to be said for gracefully finishing existing requests while new requests get routed to the new system. There are plenty of external systems outside our control that we don't want to break every time we push an update, for example.

> We have things that fall over horribly in Vumi when postgresql restarts simply because that particular piece was designed with the assumption that stuff can get restarted at any point in time. The reality is that that happens all the time and clients should handle those situations.

True, but there's a difference between handling a situation in exceptional circumstances and expecting it regularly. I've had to write code that retries something a bunch of times in a tight loop until it succeeds, and I don't want to have to do it again.

**Comment by smn** [comment]
> I disagree. There's a lot to be said for gracefully finishing existing requests while new requests get routed to the new system. There are plenty of external systems outside our control that we don't want to break every time we push an update, for example.



We have things that fall over horribly in Vumi when postgresql restarts simply because that particular piece was designed with the assumption that stuff can get restarted at any point in time. The reality is that that happens all the time and clients should handle those situations.
True, but there's a difference between handling a situation in exceptional circumstances and expecting it regularly. I've had to write code that retries something a bunch of times in a tight loop until it succeeds, and I don't want to have to do it again.

**Comment by smn** [comment]
Oops :) Reopened.

**Comment by smn** [comment]
Perhaps there's value in allowing a per account HAProxy for routing internal services? Then implications for a config change would be isolated. Really just thinking out loud atm. I don't like us making the assumption that all services will be HTTP based, that seems short sighted.

**Comment by jerith** [comment]
There's nothing stopping us from using nginx for HTTP things and haproxy for non-HTTP things.

**Comment by JayH5** [comment]
I don't have much of an understanding of how much of A Bad Thing it would be to have connections dropped but I'm likely to trust Jeremy on this one.

From what I can see, that nginx module is in the free version, but... `The ngx_stream_core_module module is available since version 1.9.0. This module is not built by default, it should be enabled with the --with-stream configuration parameter.` ... which sounds like we'd have to compile nginx ourselves? Even if we got that working, there are a bunch of useful things in HAProxy that are paid features in nginx, like basically any of the options when load balancing across upstream services (`max_conns` for example).

It is technically possible to do a smooth reload of HAProxy. Yelp have done it [here](http://engineeringblog.yelp.com/2015/04/true-zero-downtime-haproxy-reloads.html), although they don't seem to have open sourced that. It's something that we could probably figure out how to do ourselves but would likely require a significant investment of time/effort. But may also be pretty valuable.

I'm hesitant to add any more complexity to this system. I'd rather not have both HAProxy and nginx. But I guess it's an option... @jerith would we then be ok with non-HTTP connections being dropped during restarts/upgrades?

**Comment by jerith** [comment]
To clarify: I think graceful restarts (instead of murdering unfinished requests in the face) are really important for external-facing HTTP services, and we have enough of those that it's a case worth designing for.

I'd much rather have two different solutions to two different problems running side by side than have a single compromise solution that solves both problems poorly.

**Comment by jerith** [comment]
> would we then be ok with non-HTTP connections being dropped during restarts/upgrades?

Do we have any non-HTTP connections that aren't also persistent? Do we have any non-HTTP connections to internal dependencies? (I'm excluding things like rabbitmq and postgres, because we don't update/redeploy/move those very often.)

I suspect the answer to both of my questions is "no", so I'm inclined to worry about non-HTTP things later (where "later" could be next week or something).

**Comment by smn** [comment]
Well one of the things we already expect to need is the ability to run Freeswitch. Question is whether we run that through Marathon or directly on the host. I'd prefer not to automatically default to the host for everything that doesn't fit the HTTP mold.

**Comment by JayH5** [comment]
Colin and I are pretty keen to have Freeswitch run on its own host. But Freeswitch is really a special case as it uses loads of ports and so is not really suited to the one-by-one container-to-host port mapping that we currently have.

Once we get IP-per-container networking (sometime in the distant future) it will be a lot easier to run things like Freeswitch on containers that run on random hosts that Marathon/Mesos decides on.

I agree we shouldn't default to the host for everything that doesn't fit but at the moment we're really constrained with what we can do.

**Comment by smn** [comment]
:+1: glad you've thought it through then. :cake: & :shipit:

**Comment by JayH5** [comment]
Small fix based on testing.

**Pull request merged**